keel-api-jackson/src/main/kotlin/com/netflix/spinnaker/keel/json/KeelApiModule.kt

```diff
@@ -60,7 +62,9 @@ object KeelApiModule : SimpleModule("Keel API") {
         NamedType<DebianArtifact>(ArtifactType.deb.name),
         NamedType<DockerArtifact>(ArtifactType.docker.name),
         NamedType<DockerVersioningStrategy>(ArtifactType.docker.name),
-        NamedType<DebianSemVerVersioningStrategy>(ArtifactType.deb.name)
+        NamedType<DebianSemVerVersioningStrategy>(ArtifactType.deb.name),
+        NamedType<PipelineConstraintStateAttributes>("pipeline"),
+        NamedType<CanaryConstraintAttributes>("canary")
```

Comment:
This is how we know what the subtypes are, but in order to do this they had to be defined in the keel api module. This felt fine to me. I didn't do this in the `KeelConfigurationFinalizer` like resource spec things because the subtypes aren't part of a component right now, and I would have had to make them part of a component, then collect those in a list, then iterate over them like we do [here](https://github.com/spinnaker/keel/blob/master/keel-web/src/main/kotlin/com/netflix/spinnaker/config/KeelConfigurationFinalizer.kt#L32).

Reply:
Hmm, so I thought the whole point of what @robzienert had done was to make it so that `keel-api` was the only dependency a plugin developer would need to write a new `ConstraintEvaluator`, but the moment you add implementation-specific types in `keel-api`, that signals to me that a developer outside our team would need to similarly make PRs for keel to add their constraint implementations... But I'll defer to Rob on this.

Reply:
So, agree 100% with your conclusion that this is not the best way to do things. At the moment I can't figure out how to do this in a different way (I feel like there is one though). I explored [this helper](https://github.com/spinnaker/kork/blob/master/kork-core/src/main/java/com/netflix/spinnaker/kork/jackson/ObjectMapperSubtypeConfigurer.java) but couldn't find an example of it in use.

Reply:
Good find! There are a few examples in clouddriver and kayenta: https://github.com/search?q=org%3Aspinnaker+ObjectMapperSubtypeConfigurer+-repo%3Aspinnaker%2Fkork&type=Code

Reply:
I wonder if a better option might be to revert #1196 temporarily, and ask @robzienert for guidance on how best to address the sub-type registration in a Spinnaker plugin world.

Reply:
Another idea: `ConstraintEvaluators`s are `@Component`s. We could add a method like `getConstraintStateAttributeType()` to the interface, which would return the sub-type of `ConstraintStateAttribute` used by that evaluator. Then, in `KeelConfigFinalizer`, you could iterate the list of evaluators and register those sub-types.

Reply:
Sorry for weighing in late here, but isn't the right approach to have `ConstraintEvaluator` (implementations of which will be `@Component`) require implementations to specify sub-types of `Constraint` and `ConstraintStateAttribute` much like `ResourceHandler` implementations have to specify the `ResourceSpec` type? We can then easily register these on app start.

Reply:
Yeah, in fact `ConstraintEvaluator` already does this for the constraint types themselves with the `supportedType` property. Could we not just do the same thing for the attributes?

Reply:
Yup, same suggestion I gave above, although you did a much better job describing it. 🙂

Reply:
Alright, I fixed it 🤞